File: ABC-199/E.py
import sys

sys.setrecursionlimit(10 ** 6)


def dfs(v: int, p: int = -1) -> None:
    if cnt[color[v]] == 0:
        ans.append(v + 1)
    cnt[color[v]] += 1
    for u in to[v]:
        if u == p:
            continue
        dfs(u, v)
    cnt[color[v]] -= 1


if __name__ == "__main__":
    N = int(input())
    color = [0] * N
    cnt = [0] * 100005
    to = [[] for _ in range(N)]
    ans = []

    color = list(map(int, input().split()))
    for i in range(N - 1):
        a, b = map(int, input().split())
        a -= 1
        b -= 1
        to[a].append(b)
        to[b].append(a)
    dfs(0, -1)

    ans.sort()
    print(*ans)

File: ABC-199/E.cpp
#include <iostream>
#include <vector>

#define rep(i,n) for (int i=0; i<(n);++i)
using namespace std;

int n;
vector<int> color;
// どの頂点に行けるか
vector<vector<int>> to;
// 何回か
vector<int> cnt;
vector<int> ans;


// p -> おや, 親にdfsしないような仕組み
void dfs(int v, int p=-1){
    if (cnt[color[v]]==0){
        ans.push_back(v+1);
    }
    cnt[color[v]]++;
    for(int u : to[v]){
        if(u==p)continue;
        dfs(u,v);
    }
    cnt[color[v]]--;
}
int main(){
    cin>>n;
    color = vector<int>(n);
    cnt = vector<int>(100005);
    rep(i,n) cin >> color[i];
    to.resize(n);

    rep(i,n-1){
        int a,b;
        cin >> a >> b ;
        --a;
        --b;
        to[a].push_back(b);
        to[b].push_back(a);
    }
    dfs(0);
    sort(ans.begin(),ans.end());
    for(int v: ans){
        cout<< v << endl;
    }
    return 0;
}
